drivers/i2s/i2s_cavs.c

```diff
@@ -504,92 +446,163 @@ static int i2s_cavs_configure(struct device *dev, enum i2s_dir dir,
+static int i2s_tx_stream_start(struct i2s_cavs_dev_data *dev_data,
 			   volatile struct i2s_cavs_ssp *const ssp,
 			   struct device *dev_dma)
 {
-	size_t mem_block_size;
-	int ret;
+	int ret = 0;
+	void *buffer;
+	unsigned int key;
+	struct stream *strm = &dev_data->tx;
 
-	ret = queue_get(&strm->mem_block_queue, strm->cfg.options,
-			&strm->mem_block, &mem_block_size);
-	if (ret < 0) {
+	/* retrieve buffer from input queue */
+	ret = k_msgq_get(&strm->in_queue, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		SYS_LOG_ERR("No buffer in input queue to start transmission");
+		return ret;
+	}
+
+	ret = dma_reload(dev_dma, strm->dma_channel, (u32_t)buffer,
+			(u32_t)&ssp->ssd, dev_data->cfg.block_size);
+	if (ret != 0) {
+		SYS_LOG_ERR("dma_reload failed (%d)", ret);
 		return ret;
 	}
-	k_sem_give(&strm->sem);
 
-	/* Assure cache coherency before DMA read operation */
-	DCACHE_CLEAN(strm->mem_block, mem_block_size);
+	/* put buffer in output queue */
+	ret = k_msgq_put(&strm->out_queue, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		SYS_LOG_ERR("failed to put buffer in output queue");
+		return ret;
+	}
+
+	ret = dma_start(dev_dma, strm->dma_channel);
 
-	ret = start_dma(dev_dma, strm->dma_channel, &strm->dma_cfg,
-			strm->mem_block, (void *)&(ssp->ssd),
-			mem_block_size);
 	if (ret < 0) {
-		SYS_LOG_ERR("Failed to start TX DMA transfer: %d", ret);
+		SYS_LOG_ERR("dma_start failed (%d)", ret);
 		return ret;
 	}
 
-	/* enable port */
-	ssp->ssc0 |= SSCR0_SSE;
+	/* Enable transmit operation */
+	key = irq_lock();
+	ssp->sstsa |= SSTSA_TXEN;
+	irq_unlock(key);
 
-	/* Enable DMA service request handshake logic. Though DMA is
-	 * already started, it won't work without the handshake logic.
-	 */
-	ssp->ssc1 |= SSCR1_TSRE;
-	ssp->sstsa |= (0x1 << 8);
+	return 0;
+}
+
+static int i2s_rx_stream_start(struct i2s_cavs_dev_data *dev_data,
+		volatile struct i2s_cavs_ssp *const ssp, struct device *dev_dma)
+{
+	int ret = 0;
+	void *buffer;
+	unsigned int key;
+	struct stream *strm = &dev_data->rx;
+
+	/* allocate receive buffer from SLAB */
+	ret = k_mem_slab_alloc(dev_data->cfg.mem_slab, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		SYS_LOG_ERR("buffer alloc from mem_slab failed (%d)", ret);
+		return ret;
+	}
+
+	DCACHE_INVALIDATE(buffer, dev_data->cfg.block_size);
+
+	ret = dma_reload(dev_dma, strm->dma_channel, (u32_t)&ssp->ssd,
+			(u32_t)buffer, dev_data->cfg.block_size);
+	if (ret != 0) {
+		SYS_LOG_ERR("dma_reload failed (%d)", ret);
+		return ret;
+	}
+
+	/* put buffer in input queue */
+	ret = k_msgq_put(&strm->in_queue, &buffer, K_NO_WAIT);
+	if (ret != 0) {
+		SYS_LOG_ERR("failed to put buffer in output queue");
+		return ret;
+	}
+
+	SYS_LOG_INF("Starting DMA Ch%u", strm->dma_channel);
+	ret = dma_start(dev_dma, strm->dma_channel);
+	if (ret < 0) {
+		SYS_LOG_ERR("Failed to start DMA Ch%d (%d)", strm->dma_channel,
+				ret);
+		return ret;
+	}
+
+	/* Enable Receive operation */
+	key = irq_lock();
+	ssp->ssrsa |= SSRSA_RXEN;
+	irq_unlock(key);
 
```

Comment:
The sequence is l

> Most part of this function is similar to i2s_dma_rx_callback. Is it possible to merge the two?

It's possible. however since the sequence is slightly different and the hardware flags that are set are different, I kept them separate.